packages/pigeon/lib/java_generator.dart

```diff
@@ -765,22 +771,48 @@ Result<$returnType> $resultName =
     });
   }
 
+  void _writeErrorClass(Indent indent) {
+    indent.write('public static class ApiError extends RuntimeException ');
+    indent.addScoped('{', '}', () {
+      indent.writeln('public final String code;');
+      indent.writeln('public final String message;');
+      indent.writeln('public final Object details;');
+      indent.newln();
+      indent.writeln(
+          'public ApiError(@NonNull String code, @Nullable String message, @Nullable Object details) ');
+      indent.writeScoped('{', '}', () {
+        indent.writeln('super(message);');
+        indent.writeln('this.code = code;');
+        indent.writeln('this.message = message;');
+        indent.writeln('this.details = details;');
+      });
+    });
+  }
+
   void _writeWrapError(Indent indent) {
     indent.format('''
 @NonNull
 private static ArrayList<Object> wrapError(@NonNull Throwable exception) {
 \tArrayList<Object> errorList = new ArrayList<Object>(3);
-\terrorList.add(exception.toString());
-\terrorList.add(exception.getClass().getSimpleName());
-\terrorList.add(
-\t\t"Cause: " + exception.getCause() + ", Stacktrace: " + Log.getStackTraceString(exception));
+\tif (exception instanceof ApiError) {
+\t\tApiError error = (ApiError) exception;
+\t\terrorList.add(error.code);
+\t\terrorList.add(error.message);
+\t\terrorList.add(error.details);
+\t} else {
+\t\terrorList.add(exception.getClass().getSimpleName());
+\t\terrorList.add(exception.toString());
+\t\terrorList.add(
```

Comment:
This change makes sense, but I think we should keep the original order of these list items. This would be a breaking change with no "real" benefit.